riff

```diff
@@ -133,24 +211,267 @@ Usage:
+  target=$2
+  common=$source
+  result=
+  while [ "${target#"$common"}" = "$target" ]; do
+      if [ -z "$result" ] ; then
+          result=$(basename "$common")
+      else
+          result=$(basename "$common")/$result
+      fi
+      common=$(dirname "$common")
+  done
+  if [[ "$common" == "/" ]] ; then
+    result="/"$result
+  fi
+  echo $result
+}
+
+function write_dockerfile() {
+  if [[ $FNLANG == "shell" ]] ; then
+    cat >> $FNDOCKER <<EOF
+FROM projectriff/${FNLANG}-function-invoker:${RIFF_VERSION}
+ARG FUNCTION_URI="/${FNFILE}"
+ENV FUNCTION_URI \${FUNCTION_URI}
+ADD ${ARTRELPATH} /
+EOF
+  elif [[ $FNLANG == "java" ]] ; then
+    cat >> $FNDOCKER <<EOF
+FROM projectriff/${FNLANG}-function-invoker:${RIFF_VERSION}
+ARG FUNCTION_JAR=/functions/${FNFILE}
+ARG FUNCTION_CLASS=${FNCLASS}
+ENV FUNCTION_URI file://\${FUNCTION_JAR}?handler=\${FUNCTION_CLASS}
+ADD ${ARTRELPATH} \${FUNCTION_JAR}
+EOF
+  elif [[ $FNLANG == "node" ]] ; then
+    cat >> $FNDOCKER <<EOF
+FROM projectriff/${FNLANG}-function-invoker:${RIFF_VERSION}
+ENV FUNCTION_URI /functions/${FNFILE}
+ADD ${ARTRELPATH} \${FUNCTION_URI}
+EOF
+  elif [[ $FNLANG == "python" ]] ; then
+    cat >> $FNDOCKER <<EOF
+FROM projectriff/python2-function-invoker:${RIFF_VERSION}
+ARG FUNCTION_MODULE=${FNFILE}
+ARG FUNCTION_HANDLER=process
+ENV FUNCTION_URI file:///\${FUNCTION_MODULE}?handler=\${FUNCTION_HANDLER}
+ADD ${ARTRELPATH} /
+ADD ${ARTRELDIR}/requirements.txt /
+RUN pip install --upgrade pip && pip install -r /requirements.txt
+EOF
+  fi
+}
+
+function write_function_yaml() {
+  cat >> $FNDEF <<EOF
+apiVersion: projectriff.io/v1
+kind: Function
+metadata:
+  name: ${FUNCTION}
+spec:
+  protocol: ${FNPROT}
+  input: ${TOPIC_IN}
+EOF
+  if [[ -n "$TOPIC_OUT" ]] ; then
+    cat >> $FNDEF <<EOF
+  output: ${TOPIC_OUT}
+EOF
+  fi
+  cat >> $FNDEF <<EOF
+  container:
+    image: ${USERACCT}/${FUNCTION}:${VERSION}
+EOF
+}
+
+function write_topics_yaml() {
+  cat >> $FNTOPICS <<EOF
+apiVersion: projectriff.io/v1
+kind: Topic
+metadata:
+  name: $TOPIC_IN
+spec:
+  partitions: 1
+EOF
+  if [[ -n "$TOPIC_OUT" ]] ; then
+    cat >> $FNTOPICS <<EOF
+---
+apiVersion: projectriff.io/v1
+kind: Topic
+metadata:
+  name: $TOPIC_OUT
+spec:
+  partitions: 1
+EOF
+  fi
+}
+
+function build() {
+  echo "Building $FUNCTION $FNPATH $VERSION"
+  docker build -t "${USERACCT}/${FUNCTION}:${VERSION}" $FNPATH
+  if [[ $DOCKERPUSH == "true" ]] ; then
+      docker push "${USERACCT}/${FUNCTION}:${VERSION}"
+  fi
 }
 
 function apply() {
   echo "Applying $FNPATH resource[s]"
   kubectl apply -f $FNPATH
 }
 
+function create() {
+  echo "Creating function $FUNCTION version $VERSION"
+  init
+  if [ -f "${FNPATH}" ] ; then
+    FNPATH=$(dirname "$FNPATH")
+  fi
+  build
+  apply
+}
+
+function update() {
+  if [ -f "${FNPATH}" ] ; then
+    FNPATH=$(dirname "$FNPATH")
+  fi
+  FNDOCKER="$FNPATH/Dockerfile"
+  if [ -f $FNDOCKER ] ; then
+    echo "Updating $FNPATH resource[s]"
+    kubectl delete -f $FNPATH
```

Comment:
shouldn't we leave the topics alone?

Reply:
how do you change partition count? seems like our controllers don't apply what changed for modified resources

Reply:
also, if we don't delete the topic I get the `The Topic "hello" is invalid: apiVersion: Invalid value: "projectriff.io/__internal": must be projectriff.io/v1` error

Reply:
AFAIK that's an open issue for topic-controller (perhaps this one: https://github.com/projectriff/topic-controller/issues/3)

@ericbottard can clarify

One thing to keep in mind is that topics can be shared across multiple functions, so we'll need to decide how to handle that. I guess the proper behavior for `apply` with a topic resource YAML file would be to only consider changes (e.g. if the partition count is different than the existing topic). If multiple functions share a topic, and they have the same generated YAML for that topic, it wouldn't be an issue. But we might want to consider whether we should provide another flag for `init` so that topic YAML need not be generated. With a more sophisticated CLI at some point, we could actually add generate-if-not-exists functionality.

Reply:
Considering topic resource deletions are not yet processed by `topic-controller` (as you reported via https://github.com/projectriff/topic-controller/issues/17), I guess we might as well go ahead with this one as-is. We can revisit once that work is completed for the `topic-controller`.